panther_hardware_interfaces/src/roboteq_driver.cpp

```diff
@@ -100,110 +98,84 @@ bool RoboteqDriver::WaitForBoot()
-                                    [RoboteqCANObjects::fault_script_flags.subid]),
-    2);
+  std::unique_lock<std::mutex> lck_p(position_timestamp_mtx_);
+  states.pos_timestamp = last_position_timestamp_;
 
-  fb.runtime_stat_flag_motor_1 = GetByte(
-    static_cast<int32_t>(
-      rpdo_mapped[RoboteqCANObjects::motor_flags.id][RoboteqCANObjects::motor_flags.subid]),
-    0);
-  fb.runtime_stat_flag_motor_2 = GetByte(
-    static_cast<int32_t>(
-      rpdo_mapped[RoboteqCANObjects::motor_flags.id][RoboteqCANObjects::motor_flags.subid]),
-    1);
+  std::unique_lock<std::mutex> lck_sc(speed_current_timestamp_mtx_);
+  states.vel_current_timestamp = last_speed_current_timestamp_;
 
-  std::unique_lock<std::mutex> lck(rpdo_timestamp_mtx_);
-  fb.timestamp = last_rpdo_write_timestamp_;
-
-  return fb;
+  return states;
 }
 
-// todo check what happens when publishing is stopped (on hold - waiting for decision on changing to
-// PDO)
-void RoboteqDriver::SendRoboteqCmdChannel1(const std::int32_t cmd)
+RoboteqDriverState RoboteqDriver::ReadRoboteqDriverState()
 {
-  try {
-    SyncSDOWrite<std::int32_t>(RoboteqCANObjects::cmd_1.id, RoboteqCANObjects::cmd_1.subid, cmd);
-  } catch (const std::runtime_error & e) {
-    throw std::runtime_error(
-      "Error when trying to send channel 1 Roboteq command: " + std::string(e.what()));
-  }
+  RoboteqDriverState state;
+
+  state.fault_flags = GetByte(
+    static_cast<std::int32_t>(
+      rpdo_mapped[RoboteqCANObjects::flags.id][RoboteqCANObjects::flags.subid]),
+    0);
+  state.runtime_stat_flag_motor_1 = GetByte(
+    static_cast<std::int32_t>(
+      rpdo_mapped[RoboteqCANObjects::flags.id][RoboteqCANObjects::flags.subid]),
+    1);
+  state.runtime_stat_flag_motor_2 = GetByte(
+    static_cast<std::int32_t>(
+      rpdo_mapped[RoboteqCANObjects::flags.id][RoboteqCANObjects::flags.subid]),
+    2);
+  state.script_flags = GetByte(
+    static_cast<std::int32_t>(
+      rpdo_mapped[RoboteqCANObjects::flags.id][RoboteqCANObjects::flags.subid]),
+    3);
+
+  state.mcu_temp = rpdo_mapped[RoboteqCANObjects::mcu_temp.id][RoboteqCANObjects::mcu_temp.subid];
+  state.heatsink_temp =
+    rpdo_mapped[RoboteqCANObjects::heatsink_temp.id][RoboteqCANObjects::heatsink_temp.subid];
+  state.battery_voltage =
+    rpdo_mapped[RoboteqCANObjects::battery_voltage.id][RoboteqCANObjects::battery_voltage.subid];
+  state.battery_current_1 = rpdo_mapped[RoboteqCANObjects::battery_current_1.id]
+                                       [RoboteqCANObjects::battery_current_1.subid];
+  state.battery_current_2 = rpdo_mapped[RoboteqCANObjects::battery_current_2.id]
+                                       [RoboteqCANObjects::battery_current_2.subid];
+
+  std::unique_lock<std::mutex> lck_fa(flags_current_timestamp_mtx_);
+  state.flags_current_timestamp = flags_current_timestamp_;
+
+  std::unique_lock<std::mutex> lck_vt(voltages_temps_timestamp_mtx_);
+  state.voltages_temps_timestamp = last_voltages_temps_timestamp_;
```

Comment:
Same here. What about unlocking it or even using `scoped_lock`?

Reply:
as above